### PR TITLE
[IM] Add Fabric Filtered Write Support

### LINKED ISSUE #12053: Fabric Filtered Reads/Writes/Subscribes
Implemented fabric-filtered reads/writes/subscribes.

This should likely happen within the `AttributeAccessInterface`, very similar to what is being done for chunking in #12019, i.e the application/cluster handler shouldn't have to manage this bit of logic. Rather, the IM can automatically transpose the list indices as needed.

### PR DESCRIPTION
**Pull request by erjiaqing**
#### Problem
Fabric filtered write is not well supported.

Fixes #12053

#### Change overview
Add support to fabric filtered write
- Updated Decode() function (accepts a optional fabric index, for overwriting the fabric index in TLV)
- Add DecodeWithoutFabricIndex() for client side decoding.
- Update ACL cluster server

#### Testing
WIP, please review the logic first

### CODE REVIEW

File: src/app/AttributeAccessInterface.h
Comment (bzbarsky-apple) [inline src/app/AttributeAccessInterface.h:335]:
```suggestion
        return DataModel::Decode(mReader, aArg, MakeOptional(mAccessingFabricIndex));
```

File: src/app/data-model/DecodableList.h
Comment (bzbarsky-apple) [inline src/app/data-model/DecodableList.h:149]: I'm not sure I understand something, and this line highlights it.  We already have a way to represent "no fabric": it's `kUndefinedFabricIndex`.  Why do we need an `Optional<FabricIndex>` when calling Decode?

File: src/app/data-model/DecodableList.h
Comment (bzbarsky-apple) [inline src/app/data-model/DecodableList.h:209]: When would this be used?  Needs documentation.

File: src/app/data-model/Decode.h
Comment (bzbarsky-apple) [inline src/app/data-model/Decode.h:113]: No, that is not in fact the signature this is testing for.

File: src/app/data-model/DecodeWithoutFabricIndex.h
Comment (bzbarsky-apple) [inline src/app/data-model/DecodeWithoutFabricIndex.h:34]: When would this be used?  What does it mean to use it?  Needs _very_ careful documentation.

File: src/app/data-model/Decode.h
Comment (bzbarsky-apple) [inline src/app/data-model/Decode.h:122]: Is it "overwrite" as here or "override" as in other places in this PR?

And more to the point: what are the semantics of that Optional argument?  What does it actually do?  Needs to be documented.

File: src/app/clusters/access-control-server/access-control-server.cpp
Comment (mlepage-google) [inline src/app/clusters/access-control-server/access-control-server.cpp:227]: Hi Song, just curious, is this all that needs to happen for this file (access control cluster server)? How can I check how this actually works, to ensure it handles the case of:

- entire list comes in for write acl attribute
- we transcode entire list into Access Control entries and write/update them in AccessControl module

Does it somehow "skip" any non-fabric entries? (If so, I need to ensure the index we keep track of during iteration is correct...)

Comment (erjiaqing) [reply src/app/clusters/access-control-server/access-control-server.cpp:227]: IIUC, It does not skip the non-fabric entries, since the decoder knows nothing about how the acl entries are stored actually.

Comment (mlepage-google) [reply src/app/clusters/access-control-server/access-control-server.cpp:227]: I think @mrjerryjohns was saying we only do the fabric-filtered write, so on the client side, we only ever send the fabric-filtered list. I just don't know what's going to come out on this side, all entries, or just fabric-filtered. Ideally, just fabric-filtered. But you know what, it doesn't matter right now, when this PR goes in, I will be pulling it and running it and I'll make it work here either way, my module code can handle either.

Comment (mrjerryjohns) [reply src/app/clusters/access-control-server/access-control-server.cpp:227]: This will only give you the fabric filtered entries, not the entire list of entries across all fabrics. So you'll have to manage that somehow..

File: src/app/data-model/Decode.h
Comment (mrjerryjohns) [reply src/app/data-model/Decode.h:121]: But why is it optional though?

File: src/app/data-model/DecodeWithoutFabricIndex.h
Comment (mrjerryjohns) [inline src/app/data-model/DecodeWithoutFabricIndex.h:42]: I'd very much like to avoid this duality of Decode functions just to tip toe around over-writing the fabric index issue.

Why not just have a single Decode signature as before, that would attempt to decode the fabric index if present in the payload but it not present, would keep the field value to its default.

There-after, you can call a method on it to over-write/set the fabric index in the specific code paths that necessitate that.

File: src/app/clusters/access-control-server/access-control-server.cpp
Comment (mrjerryjohns) [reply src/app/clusters/access-control-server/access-control-server.cpp:227]: This logic here needs to be fixed for this PR to work correctly. Since ACLs are the target for a lot of this logic, it would make sense to show-case that being fixed up in this PR itself.

**Review by mrjerryjohns — CHANGES_REQUESTED** [review]

File: src/app/data-model/DecodeWithoutFabricIndex.h
Comment (erjiaqing) [reply src/app/data-model/DecodeWithoutFabricIndex.h:42]: The users should be aware when not overwritting the fabric index field.
For example, for ACL clusters, if we called Decode without a fabric index, the client might be able to write an entry to a fabric it does not own. Similar to some fabric senstive values.

However, on read side, we will never override such field, this function will provide a unique API for both fabric scoped and non-fabric scoped structs.

File: src/app/clusters/access-control-server/access-control-server.cpp
Comment (erjiaqing) [reply src/app/clusters/access-control-server/access-control-server.cpp:227]: Added a commit for adding fabricIndex argument to `UpdateEntry` functions, @mlepage-google please verify if the modified logic makes sense.
(I will do another update after write chunking PR)

**Comment by woody-apple** [comment]
Fast tracking, given > 3 business days for review.

File: src/app/clusters/access-control-server/access-control-server.cpp
Comment (mlepage-google) [inline src/app/clusters/access-control-server/access-control-server.cpp:366]: allCount isn't initialized here?

**Review by mlepage-google — APPROVED** [review]
Yes this is a change I was going to make, needs to happen for fabric filtered writing. It looks like you have a Verify call in the wrong place, please fix that before merging.

**Review by mlepage-google — COMMENTED** [review]
Sorry, for me approve means for the access-control-server.cpp bits, I'll defer to other IM/SDK experts for the rest of this PR.

File: src/app/clusters/access-control-server/access-control-server.cpp
Comment (mlepage-google) [reply src/app/clusters/access-control-server/access-control-server.cpp:227]: Yes, I was going to make that small change (using the accessing fabric index when indexing into ACL attribute) after this went in, I wouldn't hold up this PR for it. The sooner this PR goes in, the sooner I can start integration running and making sure everything is working well and there are some improvements I would like to make which are easier with a running system.

File: src/app/clusters/access-control-server/access-control-server.cpp
Comment (erjiaqing) [reply src/app/clusters/access-control-server/access-control-server.cpp]: Yes, moved to below

File: src/app/data-model/DecodeWithoutFabricIndex.h
Comment (mrjerryjohns) [reply src/app/data-model/DecodeWithoutFabricIndex.h:42]: > if we called Decode without a fabric index, the client might be able to write an entry to a fabric it does not own

But server logic isn't _expected_ to directly call this `Decode` right? They would go through `AttributeValueDecoder::Decode`, which would correctly stuff in the fabric index **before** handing it to cluster logic.

Comment (erjiaqing) [reply src/app/data-model/DecodeWithoutFabricIndex.h:42]: Based on this idea, I think it is ok ot provide two signatures of Decode() and remove the `DecodeWithoutFabricIndex`

File: src/app/AttributeAccessInterface.h
Comment (bzbarsky-apple) [inline src/app/AttributeAccessInterface.h:331]: We absolutely want this, per current spec proposals.  Please make sure an issue is filed tracking this.

File: src/app/data-model/Decode.h
Comment (bzbarsky-apple) [inline src/app/data-model/Decode.h:121]: I believe this is unused now.

File: src/app/data-model/Decode.h
Comment (bzbarsky-apple) [inline src/app/data-model/Decode.h:141]: This one is also unused.

File: src/app/data-model/Nullable.h
Comment (bzbarsky-apple) [inline src/app/data-model/Nullable.h:78]: I would somewhat prefer we handled this in the IsFabricScoped impl, because it's non-obvious when looking at that code that it will end up being affected by this.  That said, that can be a followup, and if we keep this here:
```suggestion
    // The only fabric-scoped objects in the spec are events and structs inside lists, and neither one can be nullable.
```

File: src/app/tests/TestAttributeValueDecoder.cpp
Comment (bzbarsky-apple) [inline src/app/tests/TestAttributeValueDecoder.cpp:140]: Why is this commented out?

File: src/controller/TypedReadCallback.h
Comment (bzbarsky-apple) [inline src/controller/TypedReadCallback.h:85]: This is wrong, because it's no longer setting `err`, so we will not call `mOnError`.... I am surprised we have no tests for this.  :(

**Review by bzbarsky-apple — CHANGES_REQUESTED** [review]

File: src/app/tests/TestAttributeValueDecoder.cpp
Comment (erjiaqing) [reply src/app/tests/TestAttributeValueDecoder.cpp]: Miscommit testing code, uncommented.

File: src/app/data-model/DecodableList.h
Comment (erjiaqing) [reply src/app/data-model/DecodableList.h:205]: Added a TODO here.

File: src/app/clusters/access-control-server/access-control-server.cpp
Comment (erjiaqing) [reply src/app/clusters/access-control-server/access-control-server.cpp:455]: Yes, it should be `while` loop.

**Review by mrjerryjohns — APPROVED** [review]

File: src/app/data-model/DecodableList.h
Comment (bzbarsky-apple) [inline src/app/data-model/DecodableList.h:167]:
```suggestion
        // TODO: Consider some setup where this field does not exist when T is not a fabric scoped struct.
```

File: src/app/data-model/Decode.h
Comment (bzbarsky-apple) [inline src/app/data-model/Decode.h:141]: Why is this conditional enable needed?  If there is no such Decode, the code below just won't compile, right?

File: src/app/data-model/Decode.h
Comment (bzbarsky-apple) [inline src/app/data-model/Decode.h:157]: Again, why is this needed?

File: src/app/AttributeAccessInterface.h
Comment (bzbarsky-apple) [reply src/app/AttributeAccessInterface.h:331]: https://github.com/project-chip/connectedhomeip/issues/14586

File: src/app/data-model/Decode.h
Comment (erjiaqing) [reply src/app/data-model/Decode.h:141]: No, compiler will complain Decode is ambiguous for other types since there are more than one Decode() available.

Comment (erjiaqing) [reply src/app/data-model/Decode.h:141]:
```
../../zzz_generated/app-common/app-common/zap-generated/cluster-objects.cpp:25470:47: error: call of overloaded 'Decode(chip::TLV::TLVReader&, chip::app::Clus
ters::ElectricalMeasurement::Attributes::ClusterRevision::TypeInfo::DecodableType&)' is ambiguous
25470 |         ReturnErrorOnFailure(DataModel::Decode(reader, clusterRevision));
      |                              ~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~
../../src/lib/support/CodeUtils.h:191:23: note: in definition of macro 'ReturnErrorOnFailure'
  191 |         auto __err = (expr);                                                                                                       \
      |                       ^~~~
In file included from ../../src/app/data-model/DecodableList.h:21,
                 from ../../zzz_generated/app-common/app-common/zap-generated/cluster-objects.h:28,
                 from ../../zzz_generated/app-common/app-common/zap-generated/cluster-objects.cpp:20:
../../src/app/data-model/Decode.h:36:12: note: candidate: 'CHIP_ERROR chip::app::DataModel::Decode(chip::TLV::TLVReader&, X&) [with X = short unsigned int; ty
pename std::enable_if<std::is_integral<_Tp>::value, int>::type <anonymous> = 0; CHIP_ERROR = chip::ChipError]'
   36 | CHIP_ERROR Decode(TLV::TLVReader & reader, X & x)
      |            ^~~~~~
../../src/app/data-model/Decode.h:97:12: note: candidate: 'CHIP_ERROR chip::app::DataModel::Decode(chip::TLV::TLVReader&, X&) [with X = short unsigned int; CH
IP_ERROR = chip::ChipError]'
   97 | CHIP_ERROR Decode(TLV::TLVReader & reader, X & x)
      |            ^~~~~~
```

File: src/app/data-model/Decode.h
Comment (erjiaqing) [reply src/app/data-model/Decode.h:157]: The same above, we need SFINAE to gurantee only one candicate with the highest priority for every type.

File: src/app/clusters/access-control-server/access-control-server.cpp
Comment (erjiaqing) [reply src/app/clusters/access-control-server/access-control-server.cpp]: Thanks.
@mlepage-google will we have more tests against the ACL cluster?

Comment (bzbarsky-apple) [reply src/app/clusters/access-control-server/access-control-server.cpp]: This could be a YAML test right now, no?  Just try to write a single wildcard admin ACL, which should not break anything if it succeeds, and should succeed....

File: src/app/data-model/Decode.h
Comment (bzbarsky-apple) [reply src/app/data-model/Decode.h:141]: I'm not talking about the Decode on line 97, which is for cluster-object structs and does need the SFINAE.

I'm talking about the ones for Nullable/Optional, which got SFINAE added when there were different-signature Decode methods involved earlier in this PR, but now that they are not involved, we don't need the SFINAE.

Concretely, I just took it out locally and everything compiles fine.  I want those 10 minutes of my life back.  ;)

File: src/app/clusters/access-control-server/access-control-server.cpp
Comment (mlepage-google) [reply src/app/clusters/access-control-server/access-control-server.cpp]: We have pretty good unit test coverage of the underlying access control system module, and almost no coverage of the cluster itself. Yes we will have testing of the cluster, issue #10253 tracks that.

Comment (mlepage-google) [reply src/app/clusters/access-control-server/access-control-server.cpp]: Writing ACLs should work in master, I do that locally. Obviously it should work with this PR also (that's the goal).

Terminology note: this isn't what Boris meant, but to be clear for any readers, "wildcard write" is not a thing we support (at least for v1).

**Review by bzbarsky-apple — APPROVED** [review]

File: src/app/data-model/Decode.h
Comment (erjiaqing) [reply src/app/data-model/Decode.h]: Realized that I'm deleting the wrong SFINAE, thanks.

**Pull request merged**
